lightning/src/routing/router.rs

```diff
@@ -333,79 +494,226 @@ pub fn get_route<L: Deref>(our_node_id: &PublicKey, network: &NetworkGraph, targ
+						assert!(res.last().unwrap().route_hop.pubkey == *target);
+					}
+
+					// How much a path can transfer is how much the weakest link can transfer (minus fee on that link).
+					if new_entry.route_hop.pubkey == *target {
+						path_bottleneck_msat = cmp::min(path_bottleneck_msat, new_entry.available_msat);
+					} else {
+						path_bottleneck_msat = cmp::min(path_bottleneck_msat, new_entry.available_msat - new_entry.total_fee_msat);
+					}
+
+					if res.last().unwrap().route_hop.pubkey == *target {
+						break;
 					}
-				} else {
-					// We should be able to fill in features for everything except the last
-					// hop, if the last hop was provided via a BOLT 11 invoice (though we
-					// should be able to extend it further as BOLT 11 does have feature
-					// flags for the last hop node itself).
-					assert!(res.last().unwrap().pubkey == *target);
+
+					new_entry = match dist.remove(&res.last().unwrap().route_hop.pubkey) {
+						Some(payment_hop) => payment_hop,
+						None => {
+							// If we can't reach a given node, something wen't wrong during path traverse.
+							// Stop looking for more paths here because next iterations will face the exact same issue.
+							break 'paths_collection;
+						}
+					};
+					res.last_mut().unwrap().route_hop.fee_msat = new_entry.route_hop.fee_msat;
+					res.last_mut().unwrap().route_hop.cltv_expiry_delta = new_entry.route_hop.cltv_expiry_delta;
+					res.push(new_entry.clone());
 				}
-				if res.last().unwrap().pubkey == *target {
-					break;
+				res.last_mut().unwrap().route_hop.fee_msat = final_value_msat;
+				res.last_mut().unwrap().route_hop.cltv_expiry_delta = final_cltv;
+
+				let mut payment_path = PaymentPath {hops: res};
+				// Since a path allows to transfer as much sats as the smallest channel (in terms of available capacity)
+				// it has ("bottleneck"), we should recompute the fees so that other (larger) channels used in a given path
+				// don't overpay fees.
+				// This may happen because when we were selecting those channels we were not aware how much this path
+				// will transfer, and the relative fee for them might have been computed considering a larger amount.
+				payment_path.update_amount_and_recompute_fees(path_bottleneck_msat);
+
+				// Remember that we used these channels so that we don't rely on the same capacity in future paths.
+				for (idp, payment_hop) in payment_path.hops.iter().enumerate() {
+					if let Some(known_channel_amount_msat) = known_channels_amounts_available_msat.get_mut(&payment_hop.route_hop.short_channel_id) {
+						let mut used_in_channel_msat = path_bottleneck_msat + payment_hop.total_fee_msat;
+						if idp == payment_path.hops.len() - 1 {
+							// No fees on the last hop.
+							used_in_channel_msat = path_bottleneck_msat;
+						}
+						if let Some(updated_amount_msat) = known_channel_amount_msat.checked_sub(used_in_channel_msat) {
+							*known_channel_amount_msat = updated_amount_msat;
+						};
```

Comment:
else branch here is unreachable/indicates bug, no? I think you can just drop the checked_sub and let things panic if we underflow.